src/dataflow/src/render/context.rs

```diff
@@ -187,22 +188,111 @@ where
+            // Determine a key that is constrained to a literal value.
+            let constrained_key = local
+                .keys()
+                .find_map(|key| key_selector(key).map(|row| (key.clone(), row)));
+            // If we found such a key, use a more advanced implementation.
+            if let Some((key, row)) = constrained_key {
+                let (oks, errs) = &local[&key];
+
+                let result = oks
+                    .stream
+                    .unary(Pipeline, "AsCollection", move |_, _| {
+                        move |input, output| {
+                            input.for_each(|time, data| {
+                                let mut session = output.session(&time);
+                                for wrapper in data.iter() {
+                                    let batch = &wrapper;
+                                    let mut cursor = batch.cursor();
+                                    cursor.seek_key(batch, &row);
+                                    if cursor.get_key(batch) == Some(&row) {
+                                        while let Some(val) = cursor.get_val(batch) {
+                                            for datum in logic(val) {
+                                                cursor.map_times(batch, |time, diff| {
+                                                    session.give((
+                                                        datum.clone(),
+                                                        time.clone(),
+                                                        diff.clone(),
+                                                    ));
+                                                });
+                                            }
+                                            cursor.step_val(batch);
+                                        }
+                                    }
+                                }
+                            });
+                        }
+                    })
+                    .as_collection();
+
+                Some((result, errs.as_collection(|k, _v| k.clone())))
+            } else {
+                let (oks, errs) = local.values().next().expect("Empty arrangement");
+                Some((
+                    oks.flat_map_ref(move |_k, v| logic(v)),
+                    errs.as_collection(|k, _v| k.clone()),
+                ))
+            }
         } else if let Some(trace) = self.trace.get(relation_expr) {
-            let (_id, oks, errs) = trace.values().next().expect("Empty arrangement");
-            Some((
-                // oks.as_collection(|_k, v| v.clone()),
-                oks.flat_map_ref(move |_k, v| logic(v)),
-                errs.as_collection(|k, _v| k.clone()),
-            ))
+            // Determine a key that is constrained to a literal value.
+            let constrained_key = trace
+                .keys()
+                .find_map(|key| key_selector(key).map(|row| (key.clone(), row)));
```

Comment:
Could the `find_map` be changed to finding the set of constrained keys of max width?

Reply:
Yes, that's a great point. I didn't have a good sense for how to optimize this (max width seems like a fine start, though there could be reasons something else would be better). Probably if we know that one column is a unique key that would be great to use.

I can introduce `TODO` content there and file an issue. Do you have a specific use in mind that would need a specific rule at this point?

Reply:
I put in the logic that prefers keys by their length, and realized that threading unique key information through would be more complicated and instead filed and referenced an issue for that.